### PR TITLE
feat(authentication): passwordless login

### PR DESCRIPTION
**Pull request by SotiriaSte**
**What kind of change does this PR introduce?** 

- [ ] Bugfix
- [X] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other (please describe)

**Does this PR introduce a breaking change?** 

- [ ] Yes
- [X] No

**The PR fulfills these requirements:**

- [x] It's submitted to the `main` branch
- [ ] When resolving a specific issue, it's referenced in the PR's description  (e.g. `fix #xxx`, where "xxx" is the issue number)

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature 

**Other information:**

### CODE REVIEW

File: modules/authentication/src/config/index.ts [hunk 1]
```diff
@@ -9,6 +9,7 @@ import twitchConfig from './twitch.config';
 import slackConfig from './slack.config';
 import tokenConfig from './token.config';
 import localConfig from './local.config';
+import magiclinkConfig from './magiclink.config';
```
Comment (kon14) [inline modules/authentication/src/config/index.ts:12]: Convert magic link file/variable/function references to camel case.

File: modules/authentication/src/handlers/magic.link.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  GrpcError,
+  ParsedRouterRequest,
+  RoutingManager,
+  UnparsedRouterResponse,
+} from '@conduitplatform/grpc-sdk';
+import { Token, User } from '../models';
+import { status } from '@grpc/grpc-js';
+import { IAuthenticationStrategy } from '../interfaces/AuthenticationStrategy';
+import { TokenProvider } from './tokenProvider';
+import { MagicLinkTemplate as template } from '../templates';
```
Comment (kon14) [inline modules/authentication/src/handlers/magic.link.ts:19]: Import this as magicLinkTemplate for clarity as auth uses multiple templates.

File: modules/authentication/src/handlers/magic.link.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        email: user.email,
+        sender: 'no-reply',
+        variables: {
+          link,
+        },
+      })
+      .catch(e => {
+        ConduitGrpcSdk.Logger.error(e);
+      });
+    return 'Email send';
```
Comment (kon14) [inline modules/authentication/src/handlers/magic.link.ts:106]: Typo: `Email sent`

File: modules/authentication/src/handlers/magic.link.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  async verifyLogin(call: ParsedRouterRequest): Promise<UnparsedRouterResponse> {
+    const { verificationToken } = call.request.params.verificationToken;
+    const context = call.request.context;
+    if (isNil(context))
+      throw new GrpcError(status.UNAUTHENTICATED, 'No headers provided');
```
Comment (kon14) [inline modules/authentication/src/handlers/magic.link.ts:113]: This is vague. It's the `clientId` that's required.
But wouldn't security already handle that?

**Review by kon14 — CHANGES_REQUESTED** [review]

File: modules/authentication/src/handlers/magic.link.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  private emailModule: Email;
+  private initialized: boolean = false;
+
+  constructor(private readonly grpcSdk: ConduitGrpcSdk) {}
+
+  async validate(): Promise<boolean> {
+    const config = ConfigController.getInstance().config;
+    if (config.magic_link.enabled && this.grpcSdk.isAvailable('email')) {
+      this.emailModule = this.grpcSdk.emailProvider!;
+      await this.registerTemplates();
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/magic.link.ts:31]: If register does not succeed, init should fail

File: modules/authentication/src/handlers/magic.link.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    } else {
+      return (this.initialized = false);
+    }
+  }
+
+  async declareRoutes(routingManager: RoutingManager): Promise<void> {
+    const config = ConfigController.getInstance().config;
+    routingManager.route(
+      {
+        path: '/hook/magic-link',
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/magic.link.ts:42]: magic-link is not a hook. Change to /magic-link

File: modules/authentication/src/handlers/magic.link.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+      {
+        path: '/hook/magic-link/:verificationToken',
+        action: ConduitRouteActions.GET,
+        description: `A webhook used to verify a user who has received a magic link.`,
+        urlParams: {
+          verificationToken: ConduitString.Required,
+        },
+      },
+      new ConduitRouteReturnDefinition('VerifyMagicLinkLoginResponse', {
+        userId: ConduitString.Optional,
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/magic.link.ts:63]: align the returns with the rest of the authorize endpoints from other handlers.
i.e. userId is not returned, access tokens are both optional, message is not returned

File: modules/authentication/src/handlers/magic.link.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    return 'token send';
+  }
+
+  private async sendMagicLinkMail(user: User, token: Token) {
+    const serverConfig = await this.grpcSdk.config.get('router');
+    const url = serverConfig.hostUrl;
+
+    const result = { token, hostUrl: url };
+    const link = `${result.hostUrl}/hook/authentication/magic-link/${result.token.token}`;
+    await this.emailModule
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/magic.link.ts:95]: should not catch the error here. If an error happens the entire function should crash. The error here is critical, that's why it shouldn't be caught at this level.

File: modules/authentication/src/handlers/magic.link.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+      .catch(e => {
+        ConduitGrpcSdk.Logger.error(e);
+      });
+    return 'Email sent';
+  }
+
+  async verifyLogin(call: ParsedRouterRequest): Promise<UnparsedRouterResponse> {
+    const { verificationToken } = call.request.params.verificationToken;
+    const context = call.request.context;
+    if (isNil(context))
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/magic.link.ts:112]: remove, context isn't null

**Review by kkopanidis — CHANGES_REQUESTED** [review]

**Review by kkopanidis — APPROVED** [review]

**Pull request merged**